wasmdome/src/main.rs

```diff
@@ -103,6 +139,127 @@ NKEYs are sensitive and should be treated as secrets.
         Err("Did not obtain valid arena credentials".into())
     }
 }
+
+fn check_schedule(nc: nats::Connection) -> Result<(), Box<dyn Error>> {
+    let res = nc.request_timeout(
+        "wasmdome.public.arena.schedule",
+        "",
+        std::time::Duration::from_millis(500),
+    );
+
+    if let Err(e) = res {
+        println!("Error requesting schedule from the lattice: {}", e);
+        return Ok(());
+    };
+
+    let data = &res.unwrap().data;
+    let matches: Vec<StoredMatch> = serde_json::from_slice(data)?;
+    let mut table = Table::new();
+    if matches.len() > 0 {
+        table.add_row(row![
+            "Match Id",
+            "Match Start",
+            "Board Size (H ⨯ W)",
+            "Max Actors",
+            "Max Turns",
+            "Actions per turn"
+        ]);
+        matches.iter().for_each(|m| {
+            table.add_row(row![
+                format!("{}", m.match_id),
+                format!("{}", m.entry.match_start),
+                format!("{} ⨯ {}", m.entry.board_height, m.entry.board_width),
+                format!("{}", m.entry.max_actors),
+                format!("{}", m.entry.max_turns),
+                format!("{}", m.aps_per_turn),
+            ]);
+        });
+    } else {
+        println!("No schedule results available.");
+        return Ok(());
+    };
+
+    table.printstd();
+    Ok(())
+}
+
+fn run_match(
+    nc: nats::Connection,
+    max_turns: u32,
+    board_height: u32,
+    board_width: u32,
+    aps_per_turn: u32,
+) -> Result<(), Box<dyn Error>> {
+    // publish game start command onto local lattice (through nc)
+    // subscribe to arena events topic and then display match conclusion event, or timeout with an error
+
+    let req = nc.request_timeout(
+        "wasmdome.internal.arena.control",
+        &serde_json::to_vec(&QueryMechs)?,
+        std::time::Duration::from_millis(1500),
+    );
+
+    if req.is_err() {
+        println!(
+            "No response from engine-provider, please ensure you have an engine-provider running."
+        );
+        return Ok(());
+    };
+
+    let res: MechQueryResponse = serde_json::from_slice(&req.unwrap().data)?;
+
+    if res.mechs.len() < 1 {
+        println!(
+            "No mechs were found in the lattice. Ensure you have scheduled at least one mech."
+        );
+        return Ok(());
+    };
+
+    let match_id = Uuid::new_v4().to_string();
+
+    let cm = StartMatch(CreateMatch {
+        match_id: match_id.clone(),
+        actors: res.mechs.iter().map(|m| m.id.clone()).collect(),
+        board_height,
+        board_width,
+        max_turns,
+        aps_per_turn,
+    });
+
+    println!("{:?}", cm);
+
+    nc.request_timeout(
+        "wasmdome.internal.arena.control",
+        &serde_json::to_vec(&cm)?,
+        std::time::Duration::from_millis(500),
+    )?;
+
+    let sub = nc.subscribe(&format!("wasmdome.match.{}.events", match_id).to_string())?;
```

Comment:
You don't need to subscribe to the match event stream, you can just subscribe to the arena event stream, which contains a much easier to trap event indicating the conclusion of a match.

Reply:
Here's where the engine provider publishes the match complete event: https://github.com/wascc/wasmdome/blob/master/engine-provider/src/game_logic.rs#L119

Reply:
You can use a synchronization primitive here like a waitgroup or a channel with a timeout where you simply wait until the match ended event has flipped the switch and if it doesn't end within some timeout, then you can indicate that something went wrong.

Reply:
The thing that prevented me from using a waitgroup, or earlier I tried the park example here: https://doc.rust-lang.org/std/thread/fn.park.html#examples is because of the nats subscription object. I'd get the following error concerning sharing the object between threads:
```
error[E0277]: `std::sync::mpsc::Sender<()>` cannot be shared between threads safely
   --> src/main.rs:266:9
    |
266 |       sub.with_handler(move |msg| {
    |  _________^^^^^^^^^^^^_-
    | |         |
    | |         `std::sync::mpsc::Sender<()>` cannot be shared between threads safely
267 | |         let turn_event: MatchEvent = serde_json::from_slice(&msg.data)?;
268 | |         match turn_event {
269 | |             MatchEvent::TurnEvent {
...   |
279 | |         Ok(())
280 | |     });
    | |_____- within this `[closure@src/main.rs:266:22: 280:6 tx:std::sync::mpsc::Sender<()>]`
    |
    = help: within `[closure@src/main.rs:266:22: 280:6 tx:std::sync::mpsc::Sender<()>]`, the trait `std::marker::Sync` is not implemented for `std::sync::mpsc::Sender<()>`
    = note: required because it appears within the type `[closure@src/main.rs:266:22: 280:6 tx:std::sync::mpsc::Sender<()>]`
```

I will attempt with a waitgroup / channel and see if I can get past it

Reply:
Try using a crossbeam channel, they are both send and sync.

Reply:
crossbeam channel worked great!